src/config/enterprise.rs

```diff
@@ -253,56 +307,31 @@ pub async fn try_attach(
-    // of `max_retries`.
-    for _ in 0..datadog.max_retries {
-        select! {
-            biased;
-            Ok(SignalTo::Shutdown | SignalTo::Quit) = signal_rx.recv() => return Err(PipelinesError::Interrupt),
-            report = report_serialized_config_to_datadog(&client, &endpoint, &auth, &payload) => {
-                match report {
-                    Ok(()) => {
-                        info!(
-                            "Vector config {} successfully reported to {}.",
-                            &config_version, DATADOG_REPORTING_PRODUCT
-                        );
-                        break;
-                    }
-                    Err(err) => {
-                        error!(
-                            err = ?err.to_string(),
-                            "Could not report Vector config to {}.", DATADOG_REPORTING_PRODUCT
-                        );
-
-                        if let ReportingError::StatusCode(status) = err {
-                            // If the error is 'fatal', and the user has elected to exit on fatal errors,
-                            // return control back upstream.
-                            //
-                            // A fatal error is currently determined to be one of the following:
-                            //
-                            // - 4xx class of status code (429 excepted; treated as a retry).
-                            if status.is_client_error() && status != StatusCode::TOO_MANY_REQUESTS {
-                                if datadog.exit_on_fatal_error {
-                                    return Err(PipelinesError::FatalCouldNotReportConfig);
-                                } else {
-                                    // Otherwise, break out of the loop, to avoid hitting the same error.
-                                    break;
-                                }
-                            }
-                        }
+    select! {
+        biased;
+        Ok(SignalTo::Shutdown | SignalTo::Quit) = signal_rx.recv() => return Err(PipelinesError::Interrupt),
+        report = report_serialized_config_to_datadog(&client, &endpoint, &auth, &payload, datadog.max_retries) => {
+            match report {
+                Ok(()) => {
+                    info!(
+                        "Vector config {} successfully reported to {}.",
+                        &config_version, DATADOG_REPORTING_PRODUCT
+                    );
+                }
+                Err(err) => {
+                    error!(
+                        err = ?err.to_string(),
+                        "Could not report Vector config to {}.", DATADOG_REPORTING_PRODUCT
+                    );
+
+                    if datadog.exit_on_fatal_error {
+                        return Err(PipelinesError::FatalCouldNotReportConfig);
+                    } else {
+                        return Err(PipelinesError::CouldNotReportConfig);
```

Comment:
Previously, it seems like we'd continue setting up the rest of the reporting (`datadog_metrics`, `datadog_logs`) infrastructure even if configuration reporting failed. I'm not sure if we'd like to preserve that behavior, so happy to change this if so.

Reply:
If config reporting fails, I think we should avoid instrumentation of the internal components... otherwise it's likely Vector will be sending metrics that are billed, but are otherwise disconnected from a known configuration.

It's also a good way to prove API/app key credentials, prior to using them for reporting.